File: c/connect4/evaluation.cpp
#include "stdafx.h"
#include "evaluation.h"
#include "bits.h"

namespace evaluation
{
	std::int32_t SIDE_TO_MOVE = 10;
	std::int32_t FACTOR_3 = 18;

	std::uint64_t shifts4(Connect4* state, std::uint64_t mask, std::uint32_t direction)
	{
		mask &= mask << (2 * state->winShifts[direction]);
		mask &= state->winMasks[direction];
		mask &= mask << state->winShifts[direction];
		return mask;
	}

	std::uint64_t strech4(Connect4* state, std::uint64_t mask, std::uint32_t direction)
	{
		mask |= mask >> (2 * state->winShifts[direction]);
		mask |= mask >> state->winShifts[direction];
		return mask;
	}

	std::int32_t evaluate(Connect4* state)
	{
		std::uint32_t current_player = state->activePlayer();
		std::uint32_t opponent = state->opponent();
		std::uint64_t own_tokens = state->playerTokens[current_player];
		std::uint64_t opponent_tokens = state->playerTokens[opponent];

		std::int32_t score = SIDE_TO_MOVE;
		for (int i = 0; i < 4; i++)
		{
			std::uint64_t win_positions = shifts4(state, state->fullBoard & ~opponent_tokens, i);
			std::uint64_t win_fields = strech4(state, win_positions, i);
			std::uint64_t tokens = own_tokens;
			tokens &= tokens << state->winShifts[i];
			score += bits::bitCount(win_fields & tokens);
			tokens &= tokens << state->winShifts[i];
			score += FACTOR_3 * bits::bitCount(win_fields & tokens);

			win_positions = shifts4(state, state->fullBoard & ~own_tokens, i);
			win_fields = strech4(state, win_positions, i);
			tokens = opponent_tokens;
			tokens &= tokens << state->winShifts[i];
			score -= bits::bitCount(win_fields & tokens);
			tokens &= tokens << state->winShifts[i];
			score -= FACTOR_3 * bits::bitCount(win_fields & tokens);
		}
		return score;
	}
}

File: python/evaluation.py
from connect4 import Connect4
from bits import sparse_bit_count

SIDE_TO_MOVE = 10
FACTOR_3 = 18

LOSS_SCORE = -10000
    
def shifts4(state, mask, direction):
	mask &= mask << (2 * direction)
	mask &= mask << direction
	return mask
    
def strech4(state, mask, direction):
	mask |= mask >> (2 * direction)
	mask |= mask >> direction
	return mask

def evaluate(state):
	score = SIDE_TO_MOVE
	score += direction(state, state._up)
	score += direction(state, state._right)
	score += direction(state, state._right_up)
	score += direction(state, state._right_down)
	return score

def direction(state, i):
	current_player = state.current_player()
	opponent = state.opponent()
	own_tokens = state._player_tokens[current_player]
	opponent_tokens = state._player_tokens[opponent]
	
	score = 0
	win_positions = shifts4(state, state._full_board & ~opponent_tokens, i)
	win_fields = strech4(state, win_positions, i)
	tokens = own_tokens
	tokens &= tokens << i
	score += sparse_bit_count(win_fields & tokens)
	tokens &= tokens << i
	score += FACTOR_3 * sparse_bit_count(win_fields & tokens)
           
	win_positions = shifts4(state, state._full_board & ~own_tokens, i)
	win_fields = strech4(state, win_positions, i)
	tokens = opponent_tokens
	tokens &= tokens << i
	score -= sparse_bit_count(win_fields & tokens)
	tokens &= tokens << i
	score -= FACTOR_3 * sparse_bit_count(win_fields & tokens)
	return score
